File: daily/2026/3691_Maximum_Total_Subarray_Value_II.cpp
#include <iostream>
#include <vector>
#include <map>

using namespace std;

class SegmentTree{
public:
    SegmentTree(vector<int>& nums){
        n = 1;
        while(n < nums.size()) n <<= 1;
        tMax.reserve(n * 2);
        tMin.reserve(n * 2);
        for (int i = 0; i < nums.size(); i++){
            tMax[n + i] = nums[i];
            tMin[n + i] = nums[i];
        }
        for (int i = n - 1; i > 0; i--){
            tMin[i] = min(tMin[i * 2], tMin[i * 2 + 1]);
            tMax[i] = max(tMax[i * 2], tMax[i * 2 + 1]);
        }
    }
    int query(int l, int r){
        int M = -1;
        int m = 1e9 + 7;
        l += n;
        r += n;
        while(l <= r){
            m = min(m, tMin[l]);
            M = max(M, tMax[l]);
            l = (l + 1) / 2;
            m = min(m, tMin[r]);
            M = max(M, tMax[r]);
            r = (r - 1) / 2;
        }
        return M - m;
    }
private:
    int n;
    vector<int> tMax;
    vector<int> tMin;
};

class Solution {
public:
    long long maxTotalValue(vector<int>& nums, int k) {
        int n = nums.size();
        SegmentTree T(nums);
        priority_queue<pair<int,pair<int,int>>, vector<pair<int,pair<int,int>>>, less<pair<int,pair<int,int>>>> pq;
        for (int l = 0; l < n; l++){
            pq.push({T.query(l, n - 1), {l, n - 1}});
        }
        long long ans = 0;
        for (int i = 0; i < k; i++){
            auto [v, p] = pq.top();
            pq.pop();
            auto [l, r] = p;
            ans += v;
            if (r > l) pq.push({T.query(l, r - 1), {l, r - 1}});
        }
        return ans;
    }
};

int main(){
    vector<int> nums = {1,3,2};
    int k = 2;
    // vector<int> nums = {4,2,5,1};
    // int k = 3;
    Solution S;
    cout << S.maxTotalValue(nums, k) << endl;
    return 0;
}

File: daily_python/3691_Maximum_Total_Subarray_Value_II.py
import os
import sys
from typing import List
from heapq import heapify, heappop, heappush

class Solution:
    def maxTotalValue(self, nums: List[int], k: int) -> int:
        n = len(nums)
        b = n.bit_length() - 1
        stMin = [[0] * (b + 1) for _ in range(n)]
        stMax = [[0] * (b + 1) for _ in range(n)]
        for start in range(n):
            stMin[start][0] = nums[start]
            stMax[start][0] = nums[start]
        for l in range(1, b + 1):
            for start in range(n):
                if start + (1 << (l - 1)) >= n:
                    continue
                stMin[start][l] = min(stMin[start][l-1], stMin[start + (1 << (l - 1))][l-1])
                stMax[start][l] = max(stMax[start][l-1], stMax[start + (1 << (l - 1))][l-1])
        def query(l: int, r: int) -> int:
            j = (r - l + 1).bit_length() - 1
            m = min(stMin[l][j], stMin[r - (1 << j) + 1][j])
            M = max(stMax[l][j], stMax[r - (1 << j) + 1][j])
            return M - m
        pq = [(-query(i, n - 1), i, n - 1) for i in range(n)]
        heapify(pq)
        ans = 0
        for _ in range(k):
            v, l, r = heappop(pq)
            ans += v
            if l < r:
                heappush(pq, (-query(l, r - 1), l, r - 1))
        return -ans
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            nums = list(map(int, lines[0].split('=')[1].strip()[1:-2].split(',')))
            k = int(lines[1].split('=')[1].strip()[:-1])
            dataset.append((nums, k))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for nums, k in dataset:
        results.append(solution.maxTotalValue(nums, k))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')